packages/react-reconciler/src/__tests__/ReactIncrementalErrorReplay-test.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @emails react-core
+ * @jest-environment node
+ */
+
+'use strict';
+
+let ReactFeatureFlags;
+let React;
+let ReactNoop;
+
+describe('ReactIncrementalErrorReplay', () => {
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactNoop = require('react-noop-renderer');
+  });
+
+  function div(...children) {
+    children = children.map(c => (typeof c === 'string' ? {text: c} : c));
+    return {type: 'div', children, prop: undefined};
+  }
+
+  function span(prop) {
+    return {type: 'span', children: [], prop};
+  }
+
+  it('should fail gracefully on error in the host environment', () => {
+    ReactNoop.simulateErrorInHostConfig(() => {
+      ReactNoop.render(<span />);
+      expect(() => ReactNoop.flush()).toThrow('Error in host config.');
```

Comment:
This one causes an unexpected pop because replay logic assumes we can always safely pop context after "begin" phase failed:

https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberScheduler.js#L274-L291

But this doesn't necessarily need to be the case if pushing context involves running user or renderer code:

https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberHostContext.js#L86

https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberHostContext.js#L64

In that case there's a chance we won't get to push, and the stack won't line up.

Curiously, for legacy class context we seem to avoid this problem because we first [push without running user code](https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberContext.js#L250-L255) and then ["replace" it](https://github.com/facebook/react/blob/6b99c6f9d376bacbb769264d743c405b495b03ad/packages/react-reconciler/src/ReactFiberContext.js#L283-L287). So even if user code failed, the stack would match up. Maybe we could always do something like this?

Reply:
Yeah something like this should work. Seems easy to mess up, though.

Reply:
Sounds fine, we'll have some inline comments so the next person knows what's up.